toolchain/sem_ir/name_scope.h

```diff
@@ -18,13 +18,104 @@ enum class AccessKind : int8_t {
   Private,
 };
 
+// Represents the result of a name lookup.
+// A lookup result can be:
+// * Successful returning a valid `InstId`,
+// * Not found when the name wasn't declared or poisoned.
+// * Poisoned when the name wasn't declared but was poisoned.
+// * Represent that an error has occurred during lookup. This is still
+// considered found and the error `InstId` is considered valid.
+class ScopeLookupResult {
+ public:
+  // Creates a non poisoned result.
+  // Depending on the given `inst_id`, it can be:
+  // * Found: Lookup was successful and `inst_id` is valid. If `inst_id` is
+  // known to be valid, use `CreateFound()`.
+  // * Not Found: Lookup returned no results and `inst_id` is invalid. If
+  // `inst_id` is known to be invalid, use `CreateNotFound()`.
+  // * Error: Lookup encountered an error and `inst_id` is
+  // `ErrorInst::SingletonInstId`, which is considered valid and found. If
+  // `inst_id` is known to be `ErrorInst::SingletonInstId`, `use CreateError()`.
+  //
+  // To create a poisoned result, use `CreatePoisoned()`.
+  static auto Create(InstId inst_id, AccessKind access_kind)
+      -> ScopeLookupResult {
+    return ScopeLookupResult(inst_id, access_kind, false);
+  }
+
+  // Creates a successful lookup result. `inst_id` must be valid. If
+  // `inst_id` is known to be `ErrorInst::SingletonInstId`, `use CreateError()`.
+  static auto CreateFound(InstId inst_id, AccessKind access_kind)
+      -> ScopeLookupResult {
+    CARBON_CHECK(inst_id.is_valid());
+    return Create(inst_id, access_kind);
+  }
+
+  // Creates a result when there are no result.
+  static auto CreateNotFound() -> ScopeLookupResult {
+    return Create(InstId::Invalid, AccessKind::Public);
+  }
+
+  // Creates a poisoned result.
+  static auto CreatePoisoned() -> ScopeLookupResult {
+    return ScopeLookupResult(InstId::Invalid, AccessKind::Public, true);
+  }
+
+  // Creates a result when an error was encountered, which is considered valid
+  // and found.
+  static auto CreateError() -> ScopeLookupResult {
+    return CreateFound(ErrorInst::SingletonInstId, AccessKind::Public);
+  }
+
+  // True iff CreatePoisoned() was used.
+  auto is_poisoned() const -> bool { return is_poisoned_; }
+
+  // True when lookup was successful or resulted with an error.
+  // False for poisoned, not found.
+  auto is_found() const -> bool {
+    return !is_poisoned() && lookup_inst_id().is_valid();
+  }
+
+  // The `InstId` of the result of the lookup.
+  // Must not be called for poisoned results.
+  // If valid, it means the lookup was successful or resulted with an error.
+  auto lookup_inst_id() const -> InstId {
+    CARBON_CHECK(!is_poisoned());
+    return inst_id_;
+  }
+
+  // The `InstId` of the result of the lookup.
+  // Must only be called when lookup was successful.
+  auto found_inst_id() const -> InstId {
+    CARBON_CHECK(is_found());
+    return inst_id_;
+  }
```

Comment:
When I see `lookup_inst_id` vs `found_inst_id`, I expect them to be meaningfully different values. e.g., the inst id which _caused_ a lookup, versus the _result_. What would you think of changing this to just a simple `inst_id()` accessor and doing any necessary validation separately? That looks like it could still require `!is_poisoned`.

Reply:
I agree with your concern.

There are currently 6 different use cases of `found_inst_id()`, and I expect this number to grow.
If we don't verify this internally, I can imagine that in many cases that value would not be verified.
Readers (and maintainers) of the code would have to wonder whether the `InstId` can be invalid, and are likely to try and handle all cases.
Alternatively, they might implicitly assume it's valid and if this implicit assumption is wrong it could create hard to find bugs.

Perhaps we can find better names for the functions?

Since I plan to introduce `poisoning_inst_id()` or something similar, having just `inst_id()` would probably not work well.

We could use `match_inst_id()` and `validated_match_inst_id()` (or `match_inst_id_validated()`).

WDYT?

Reply:
Looking at this closer, I think `lookup_inst_id` is used more than it needs to be... I've marked a couple spots that could be using `is_found` instead of `lookup_inst_id().is_valid()`.  If you remove it, i.e.:

```suggestion
  // The `InstId` of the result of the lookup.
  // Must only be called when lookup was successful.
  auto found_inst_id() const -> InstId {
    CARBON_CHECK(is_found());
    return inst_id_;
  }
```

and migrate every `lookup_inst_id` call to this (except for the one inside `is_found`, that needs to check `inst_id_` directly) it looks like all tests pass -- no code needs to access the value if it's not found. Perhaps it could just be deleted and everything could be using the higher level of validation?

(versus `poisoning_inst_id`, the current `lookup_inst_id` validation is a subset of `found_inst_id`, whereas `poisoning_inst_id` would presumably be divergent)

Reply:
Done.
How about naming it `matched_inst_id()`?
We could change `is_found()` to `is_matched()`.

Reply:
FWIW, for "Lookup", "found" seems better than "matched" (I'm used to "matched" more in like a regexp match). If the phrasing of "found_inst_id" sounds too query-like to you, perhaps "target_inst_id" since we use that elsewhere (i.e., the target of the lookup) or perhaps "decl_inst_id" or "named_inst_id"? Of these, "decl_inst_id" sounds slightly better to me, since a declaration should be the result of a successful name lookup.

Either way, I'd suggest keeping with `is_found` to match the `MakeFound` constructor, etc. But feel free to rename `found_inst_id` prior to merging.

Reply:
Renamed to `target_inst_id()`.
I think this contrasts the planned `poisoning_inst_id()` (or something similar) which is also an `InstId` of a declaration.